src/lib/Repository/ContentService.php

```diff
@@ -381,20 +387,41 @@ public function loadContentByVersionInfo(APIVersionInfo $versionInfo, array $lan
     public function loadContent(int $contentId, array $languages = null, ?int $versionNo = null, bool $useAlwaysAvailable = true): APIContent
     {
         $content = $this->internalLoadContentById($contentId, $languages, $versionNo, $useAlwaysAvailable);
-
         if (!$this->permissionResolver->canUser('content', 'read', $content)) {
             throw new UnauthorizedException('content', 'read', ['contentId' => $contentId]);
         }
         if (
             !$content->getVersionInfo()->isPublished()
             && !$this->permissionResolver->canUser('content', 'versionread', $content)
         ) {
-            throw new UnauthorizedException('content', 'versionread', ['contentId' => $contentId, 'versionNo' => $versionNo]);
+            if (!$this->isInGracePeriod($content, $this->settings['grace_period_in_seconds'], $versionNo)) {
+                throw new UnauthorizedException('content', 'versionread', ['contentId' => $contentId, 'versionNo' => $versionNo]);
+            } else {
+                $this->contentCollector->collectContent($content);
+            }
```

Comment:
How about an early throw to reduce nested else?
```suggestion
            if (!$this->isInGracePeriod($content, $this->settings['grace_period_in_seconds'], $versionNo)) {
                throw new UnauthorizedException('content', 'versionread', ['contentId' => $contentId, 'versionNo' => $versionNo]);
            }

            $this->contentCollector->collectContent($content);
```

Reply:
But also, I remember that in mu previous review, I've mentioned that checking if the version is archived costs less than checking for grace period. Or in the other words: if the version is not archived (but just a draft), then it's not worth to check if it's in grace period.

Reply:
And I commented on that, that here
https://github.com/ibexa/core/pull/515#discussion_r2012025933
that it dosent work with additional check, I guess content is not in proper stage there yet.

Reply:
> And I commented on that, that here [#515 (comment)](https://github.com/ibexa/core/pull/515#discussion_r2012025933) that it dosent work with additional check, I guess content is not in proper stage there yet.

Sorry, missed that response :sweat_smile: 
Looks like race condition - versionInfo was loaded into API object before it became archived indeed.
I don't have good idea how to debug it either, so I'm withdrawing the request.